### DISCUSSION TITLE
Nostr and Reticulum are swappable?

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#59] [Ideas]
A lot of people are tinkering with Nostr and building cool little apps on it:

- https://jesterui.github.io/#/faq
- https://github.com/melvincarvalho/nostr.chat
- https://github.com/fiatjaf/branle

Nostr *can* allow for e2e encryption if the client adds it, but it can *never* allow for forward secrecy. The nodes in Nostr are not “dumb” because you post data to the node and the other user goes to the node to retrieve it. 

Nostr nodes are like dead drop boxes that spies uses. And just like those, you can surveil a dead drop and wait to see who accesses the data. 

Theoretically though, LXMF and Nostr have similar goals of simplicity and such. You could probably replace the Nostr specific code with LXMF code in most of these apps and have the same result, no?

**Comment by markqvist** [comment]
Interesting, I had not seen Nostr before. In some ways, it is very similar to LXMF, in that it also has the "dead-drop" and wide distribution mechanism. It is different that LXMF offers forward secrecy and encryption by default, while still allowing the posting of public or semi-public information.

I really don't like the fact about Nostr that it knows everything about who posts what when, and that this information can be easily logged and turned against users.

But to actually answer the question ;) Yes, from what I can initially gather, you could replace Nostr with LXMF for most cases, and have the same result.

**Reply by trosel** [reply]
> it is very similar to LXMF, in that it also has the "dead-drop" and wide distribution mechanism

can you say more about this? I was wondering today, how does LXMF store messages for users long term in case they are not online right away? What about public messages? How long are they stored for?

Relays in LXMF must not be “dumb” if they can do these things, right?

**update**: Nostr only has Relays. LXMF has both “hubs” as well as “propagation nodes”. 2 layers instead of 1. Curious how that improves things

**Reply by faragher** [reply]
It's not a layer so much as a role. Any given node can route, but only some nodes store data.

Messages are stored, encrypted, on a distribution node.

**Reply by trosel** [reply]
Would it be possible for nodes to only store messages for approved clients? If that’s not possible, then nodes can’t charge money. 

So the intention is that everyone will use the node of a friend? How would friends know only friends use their relays?

**Reply by faragher** [reply]
> Would it be possible for nodes to only store messages for approved clients? If that’s not possible, then nodes can’t charge money.

I don't believe so. As it's open source software, if you believe that's important, you can add the feature either in a fork or as a pull request. However, you'll likely have to add it yourself.

> So the intention is that everyone will use the node of a friend? How would friends know only friends use their relays?

The intention is public distribution nodes will store messages for the network. Since LXMF messages are small by their nature, this results in kilobytes of storage at this time. Eventually there will have to be a feature to cull messages to abandoned addresses, but at this time a handful of distribution nodes (all but one run by people other than Mark, IIRC, simply because they want to support the network) serve the entire testnet.

There is no way to know only friends use the relay, and that's by design. Reticulum is a peer to peer communication system designed to circumvent censorship and allow people to freely communicate with each other. Restricting access is generally against that design. 

That's not to say it can't be done with some modification. NomadNet pages do have an authentication option, but here's the thing, by default, Reticulum doesn't know where a packet comes from. There's no source data unless the client goes out of its way to identify itself to a particular node. So to make sure only your friends used your node, you'd have to have them identify to your node, then write a routine to check their identity before forwarding a packet. 

If you're not interested in communicating with people outside the network, just don't connect to the public testnet. Reticulum is designed to make small networks, so by using radios or simply setting up your own TCP Server you can have a closed network where only people who connect directly can talk to each other. That being said, this is security by obscurity, as there's no way to lock out people who know the address.

Since Reticulum focuses on anonymous communication and identities, if you're looking to monetize access, it may not be the right protocol for you.

**Reply by trosel** [reply]
> if you're looking to monetize access, it may not be the right protocol for you.

Not “monetize”, but rather “support”. 

Nostr is starting to feel the pains of how much it costs to keep public relays running. Donations aren’t cutting it. 

Reticulum should be thinking ahead about this as well *before* it becomes an issue. 

What does a reticulum relay maintainer get out of it? At least in networks like the fediverse, they get a perverse feeling of what it’s like to be a dictator. But when everything is encrypted, they don’t even get *that*. The incentivization becomes even more important in a totally private network of dumb relays.

**Reply by faragher** [reply]
I run one because I had spare hardware. What others get out of it is up to them.

But you misunderstand the stack. Messages are only stored until they are sent. Reticulum is not a social network that stores posts. A distribution node is like a POP3 server or a physical mailbox: it only holds a message until it's picked up. 

Every single user connected to two interfaces is a relay. Distribution nodes temporarily hold messages for delivery. 

If someone uses Reticulum as a transfer protocol to run a message board, social media site, etc. then their servers are the ones that see the load. Distribution nodes only hold messages for LXMF. They do not store public posts, host files, or otherwise have overhead beyond the average network router.

**Reply by markqvist** [reply]
Exactly, and that makes the cost of running one completely neglible. LXMF is so efficient that it can handle millions of users on a Raspberry Pi and a 256 GB SD card. When something is that cheap, how are you realistically going to monetise it? There's no point. Leaves you to focus on the things that *are* hard to provide for monetisation.

I'm pretty sure there's already a few forum posts where I wrote a couple of pages on this topic, and I'll expand on this later, but a very general theme in everything I have implemented around Reticulum is to base reasoning on *physical resource limits* and *prioritising how they are utilised*.

Within this conceptual framework, it's possible to create functional (truly) decentralised application where cost is so distributed that it is neglible for every user. *Not* using memory resources that are available is actually more costly than giving them away for free, on a societal level.

When you have Reticulum fully deployed, there is no need for costly "servers" that centrally handle compute and memory demands. The concept is unneccesary when everyone can reach everyone. At our current level of technology, we have all the compute and memory we can possibly need, it's only the transport layer that is lacking immensely behind.

There's no point in trying to monetise the water you swim in.

**Reply by trosel** [reply]
So, as I understand it, storage will almost always happen on user devices. And for things like cloud storage (for data redundancy and recovery), that is outside the scope and concern of Reticulum (although developers will probably make use of them).

Questions remaining:

1. How long are messages stored in LXMF distribution nodes before they expire? How often do users need to log on to check if they have messages?

2.  What is the cost of this kind of server below? Surely those will need some kind of support?

> If someone uses Reticulum as a transfer protocol to run a message board, social media site, etc. then their servers are the ones that see the load.

**Reply by faragher** [reply]
> 2. What is the cost of this kind of server below? Surely those will need some kind of support?

Up front cost of about 100 USD, around 50% of which is durable. With a 6 year amortization schedule, call it $10 per year. At maximum power consumption, you'd consume 3A at 5V, or about 20 watts with an inefficient power supply. At 0.17 USD per kilowatt hour, never-exceed power costs are 176 KWh, or about 30 USD. Considering the power draw is likely 1/6 that, an estimate of 40 USD per year is a huge over-estimation, which is the way we like it.

Now, this server will barely be in use, so you can set up an NTP server, a mail server, a message board, some computer controlled lights, a trunking telephone exchange, or a weather station, or even all of the above, on the machine at the same time.

So, maximum 100 USD up front and 50 USD per year. More likely, the kind of person who would have a node already has a PiHole, print server, or other network-enabled SBC that can run on existing hardware for between 1 and 5 USD per year for power. 

Administrative costs are a thing, but hobbyists usually aren't looking for hourly wages. Now, if you want to have a DNS entry or other fancy things, those all cost more, but for me, I had a domain name for other uses, room for a VM on an existing server, and multiple SBCs for testing. So at this point, I'm paying maybe 5 USD more a year for power and everything else is existing infrastructure.

Then there's the people on the test net that run their nodes purely on solar power. I won't get into the cost of depreciation on solar panels but again, if they have extra capacity as-is, it's already paid for.

**Reply by trosel** [reply]
> Now, this server will barely be in use

How do you figure that? If it's running a social media site, it will have massive amounts of traffic and need to be able to handle it or it will be unreliable, uptime not guaranteed, and users will just stop using it because of the poor experience.

I sense some defensiveness when I suggest compensating server admins, and I don't know why. 

Again, today, most server admins are doing it because it's fun. But if your server is *dumb* (like in Reticulum), that fun factor goes down quite a lot. So what is the incentive here? 

When people start complaining "why is my blog not available", server admins will start complaining back "well pay me then!". And how will we answer that?

**Reply by faragher** [reply]
Reticulum is not a social media site. It does not host blogs. It is a network stack similar to TCP. NomadNet and Sideband are examples of apps that use the Reticulum Network Stack. 

When Facebook has an outage, nobody complains that TCP has let them down, it's Facebook's problem. When your blog goes down, that's because you're hosting it. 

Reticulum does not host content. It transfers data. The data you transfer is arbitrary. You can make a social media site that uses the RNS to transfer data, but then the actual load is on your server. The end game of Reticulum isn't to replace Twitter, it's to be in the kernel next to the TCP/IP stack. It's no more a social media server than HTML is. It's a way to request and transfer data.

If you don't see the incentive of base infrastructure work, then you may be happier developing an application that uses the RNS to connect to users, or simply use the existing TCP/IP infrastructure. I think you're expecting Reticulum to be Nostr, and it isn't. Every page you see on NomadNet is hosted locally by the node owner. In that regard, you can absolutely make fun server-side things. I've made message boards, indexing tools, and a simple echo bot. I really like the node with the weather satellite downlink data.

But all of this is on private servers linked by Reticulum. None of it is hosted by Reticulum. If you want to start your own server, NomadNet pages support Python; you can do basically whatever you want. If you're running your own server, you can absolutely restrict who can access it, as I noted before and as defined in the documentation. But that's not a "Reticulum router" it's just a server like a web server. Reticulum itself does not host any data.

**Reply by markqvist** [reply]
> When Facebook has an outage, nobody complains that TCP has let them down

This is, hands down, the best thing I've heard in a week.

This is a really interesting discussion though, and I have lots I want to add, but right now I'm just too busy putting out the releases ;)

Thanks for all the clarification and elaboration @faragher!

**Reply by trosel** [reply]
> When Facebook has an outage, nobody complains that TCP has let them down

that is because TCP never goes down. Reticulum nodes can, however. 

> You can make a social media site that uses the RNS to transfer data, but then the actual load is on your server

Server load = transferring data, no? When a server is under heavy load, it cannot keep up with the data transfer

**Reply by faragher** [reply]
> that is because TCP never goes down. Reticulum nodes can, however.

Correct, and as a mesh network, if there's an alternate route, the path will bypass the missing node. It happens all the time with people spinning up a Reticulum node and losing interest in the project. It's why there were originally two TCP servers owned by Mark and now there are at least three, only one of which is owned by Mark. If people connect to two of these servers, the network persists even if one goes down. If they connect to all three, it's largely bulletproof. Given secondary links, this becomes even more robust as we're still generally reliant on the Internet for long-range network links. It's terribly convenient, and it's not going away any time soon, but the multiple interfaces, all with a very broad concept of "interface" lead to increased resilience whenever a new physical transport layer is added.
 
> > You can make a social media site that uses the RNS to transfer data, but then the actual load is on your server
> 
> Server load = transferring data, no? When a server is under heavy load, it cannot keep up with the data transfer

Yes. However, it's important to define the scale in question. The Reticulum packet has virtually no overhead at modern speeds (at RNode speeds, it's highly efficient but noticeable at around 36 bytes, IIRC). For LXMF transport, assuming 512 bytes of packet size (using uncompressed ASCII gives you around twice the payload of a single tweet) and an average Netflix stream of 24Mbps (3MBps) you can conservatively pass 5,000 messages per second, presuming a similar upload capacity. 

If you're looking for file sharing, that consumes more bandwidth, but that's not the intent of Reticulum and, frankly, BitTorrent has a protocol far superior for bulk data transfer. Passing images and small (~10 MB) files over 115kbps links (Legacy serial) is easy peasy, and Reticulum does support arbitrary speed caps, so it depends entirely on if you're on an RNode or gigabit fiber, as well as the people in the middle and the destination.

My way of context, in 2022, there were around 6,000 tweets per second, so a single Reticulum node could pass 80% of all Twitter, assuming text posts, by itself. Given the slight delays while loading tweets from Twitter, a handful of servers could handle the load quite nicely, given some basic load leveling, likely geographically. 

So, if you're looking for an image-heavy Facebook or Twitter, then yes, fewer posts will be possible per second. However, if Twitter suddenly decided to be Reticulum only, that would lead to a rough transition, but a startup social media server would have far more problems handling 5,000 tweets per second than Reticulum would have transporting them.

It would take a server farm to overload a single Reticulum node for the same reason your home router's processor is rarely an issue: the bottleneck is the network bandwidth, not processing power. 

If you're storing images, doing database lookups, post-processing images to remove metadata and standardize their size, tracking telemetry, connecting to third party servers, running SSIs to make dynamic web pages, etc. then yes, your social media site will require extensive processing, storage, and memory capabilities. Routing packets is much, much easier.

**Reply by trosel** [reply]
Apologies for getting a bit off topic, but assuming that all processing of data can happen on the client, the only thing that would be needed to compliment Reticulum for a social media use case would be some kind of cloud storage sync, right?

reticulum for data transfer, client for data processing, ??? for data storage and multi-device data sync. I know IPFS doesn’t work fantastically well, but it seems that something like that would be common partner to reticulum, yes?

**Reply by markqvist** [reply]
One of the primary reasons I wrote Reticulum was to enable a world where nobody would need a "cloud" storage or any other kind of intermediaries to create or write applications.

Of course, anyone can build any kind of application, in any way they so choose on Reticulum. It's "just" a transport layer, so you can do whatever you want on top of it.

But for me personally, the point is to ultimately move away from the current and limited way of thinking about networked applications and services.

There is no cloud. It's just other computers people pretend don't exist.

For all practical and everyday purposes, there's enough storage, compute and bandwidth jutting around in peoples pockets to power at least 10x the current application load of the entire current internet.

**Reply by faragher** [reply]
There's quite a bit of back-end for social media that's server based instead of client based. Even if we assume all images/videos are on a third server (or really CDN) the server needs to query a database for all the posts of interest (and it would be a complicated query, looking for posts of a type, checking permissions, doing all sorts of post-processing if you don't simply want to show the most recent (such as adding weight to life events, etc.) and generally chewing up compute time. Then it has to render a page and spit it out to the client. There are many client-side scripts that handle things like infinite scrolling and real time updates, but every one of them calls home for processing on the server. 

But you're correct that using Reticulum as a transport between a traditional sever-client architecture is the way to go. 

By replacing the CDN with Reticulum resource requests will truly distribute the media, but has other plusses and minuses. It would reduce server load, but not the database lookups. Even then, the database is centralized so it's still a single-owner or lightly federated system. 

I'm not against it in any way: I love the concept of a distributed anything, and social media is so prevalent these days, so getting something up and running would be great. Being Reticulum-accessible is just a bonus.

There are a number of systems that have tried this before, from message boards or MySpace to things like Freenet (which has turned into something unsettling) so there are a lot of resources to draw from, but this is a very, very large task, and I'm not a subject matter expert. I'd talk to people from Mastadon and similar successful systems to get a basic idea of the kinds of things that are required for any social media network.

**Reply by githubques** [reply]
1) Instead of going big with a social media network to start off, how do you setup a simple html website using reticulum as the transport? Bonus if it can also utilize css and javascript/jquery.

You pointed to Mastadon as an example (bastyon might be another) of how to do a decentralized social media network likewise is there a source that explains how to step by step (assuming no one here can/has time to explain the steps) of doing a basic html site. Is Django involved?

2) If this can be done, is it possible to build a hybrid site where the html site is there as the primary site, but if something interrupted the standard internet it would fall back to a bare bones/tiny file size site optimized to run solely on RNodes?

3) It might be worth the effort by having the html reticulum sites carry a banner at the bottom with a monthly funding ticker: _"This site is powered by Reticulum, a decentralized form of the internet. Reticulum is at $450 of it's $2000 monthly funding goal with 23 days left in the month. Click here to donate."_

**Reply by trosel** [reply]
> There is no cloud. It's just other computers people pretend don't exist.

> For all practical and everyday purposes, there's enough storage, compute and bandwidth jutting around in peoples pockets to power at least 10x the current application load of the entire current internet.

@markqvist secure-scuttlebutt already tried to do the “multi device isn’t necessary” thing and people hated it. At the least you need sync between devices. And if you have that, one device will have more storage than another device. 

@faragher You will be interested in looking into how Nostr is handling this stuff. Here is an NIP for a message that is a search query: https://github.com/nostr-protocol/nips/blob/master/50.md Here is an NIP for a message that adds file metadata https://github.com/nostr-protocol/nips/blob/master/94.md

**Reply by faragher** [reply]
> 1. Instead of going big with a social media network to start off, how do you setup a simple html website using reticulum as the transport? Bonus if it can also utilize css and javascript/jquery.
> 
> 2. If this can be done, is it possible to build a hybrid site where the html site is there as the primary site, but if something interrupted the standard internet it would fall back to a bare bones/tiny file size site optimized to run solely on RNodes?

This is fundamentally the same question, so please allow me to condense them: if done from the ground up, this is easy. If you're retrofitting something, it's slightly harder.

Take Apache or NGINX solely as a web server. A user has a web browser that sends a request. The server processes the request, does any SSI, then sends the pertinent code back. Finally, the browser parses and renders the client-side code. 

This process barely involves transport. Using Reticulum is a matter of making the web browser and server talk over Reticulum. Not simple, but straightforward.

I would begin by finding an open source web server with a fairly well differentiated TCP interface, then having that respond to a Reticulum request. For practical reasons, I suggest a Python based web server. By using some example Reticulum code with different aspects, you can spoof an HTTP request and have it return some HTML. Even if you can't parse it (just make it simple!) it proves the server is active.

From there you find a browser (which is very limited in this day and age, since cores are either Gecko or Chromium, please correct me if I'm wrong) and have it send requests over Reticulum to the server's address. Data is data, so if you can simply pass back and forth over links and using resources, the work is basically done. 

The obvious question, what about SSL? Basically it's enabled by Reticulum as-is **so long as you have the server/client use Reticulum and not just patch in http->RNS as an afterthought.** Even then it's likely safe if all insecure packets are on your server, but really why would you?

Edge cases will bite you, such as sites using strange HTTP tricks, demanding HTTPS, or expecting certain latencies or throughputs, but for the base question "How do you host a web page" that's the answer.

And it also, fundamentally, means that everything the web server and browser support over TCP will, in theory, function over RNS without noticing. Streaming content may be the most delicate thing to deal with.

Which does mean a single site with a single backend can support both TCP and RNS simultaneously. If you wanted a RNS-only site, use an agent sniffer the way mobile sites work. 

Again, very briefly, were I to start this project (and I am absolutely not doing so at this time) I'd find a Python web server, get it running on RNS, then either modify an existing web browser or find a way to make a plugin. Above my pay-grade.

**Reply by githubques** [reply]
If it's "Above my pay-grade" for you, can't begin to think how far out it would be for me.

Sounds like the only way an amateur to novice web designer can create a site that runs on RNS would require someone with the skill set you mentioned to setup a web server (not just a Raspberry Pi) following the steps you outlined and then offer site hosting that novices can sign up for.

**Reply by faragher** [reply]
> If it's "Above my pay-grade" for you, can't begin to think how far out it would be for me.

While we don't know each other's skills, please don't overestimate me if it comes at the cost of your personal growth. I am not a programmer by trade, everything I know I've learned from finding a project I can't do, then learning what I don't know until the project is done. I've just been doing this for a long time and have a focus on networks (in the traditional sense), data transfer, and information storage and retrieval. My skills as a programmer are about average . . . I don't feel confident saying if I'm above or below that line.

Bottom line, if it's something that interests you, hunt it down. If it's not something that interests you, then your time is better spent pursuing a passion. 
 
> Sounds like the only way an amateur to novice web designer can create a site that runs on RNS would require someone with the skill set you mentioned to setup a web server (not just a Raspberry Pi) following the steps you outlined and then offer site hosting that novices can sign up for.

I think the Pi could handle it, but you're generally right. That's the standard model for web servers for a reason, their requirements are so specific that it's better to group a large number together in an easy to access system than run a single instance on a single computer.

However, people with static or even "sticky" IPs run their own web servers all the time, and there's no reason a LAMP stack can't add in a Reticulum server component without too much trouble. When the modified server programs are available, it would be basically seamless after installation. The problem, as always, is getting to that line. Basically it'll be enthusiast only until it's financially viable, then it'll be commercial with only enthusiasts rolling their own (but with a much lower barrier to entry).

**Reply by githubques** [reply]
_"everything I know I've learned from finding a project I can't do, then learning what I don't know until the project is done."_

Ditto, although it's mostly front end with limited php as the back end using what spare time I have available. And that's the issue, the opportunity cost is too high to jump into learning python before even getting to the point of attempting a project like an RNS web server vs current side projects.

So yeah, I'll continue to watch from the sidelines waiting for RNS development to get to the point where I can make use of it given my time/ability.

**Comment by markqvist** [comment]
> secure-scuttlebutt already tried to do the “multi device isn’t necessary” thing and people hated it. At the least you need sync between devices. And if you have that, one device will have more storage than another device.

Of course, multi-device is necessary. That's exactly one of the reasons we need a transport layer such as Reticulum. Distributed applications on the IP stack fail because it doesn't have the end-to-end features that Reticulum offers. You can't build truly decentralized applications without them, and that's why protocols such as SSB had to "simplify" their ambition with a "no multi-device" strategy. With Reticulum you don't need to do that.

**Comment by trosel** [comment]
![IMG_8586](https://github.com/markqvist/Reticulum/assets/1299322/94fa3e9e-8ad9-4103-9c02-d0a330115c03)

interesting comment from Nostr
